diff --git a/drivers/scsi/qla1280.h b/drivers/scsi/qla1280.h
index 48296a289428..ab121c2b03b4 100644
--- a/drivers/scsi/qla1280.h
+++ b/drivers/scsi/qla1280.h
@@ -1056,7 +1056,7 @@ struct scsi_qla_host {
 		uint32_t reset_active:1;		/* 3 */
 		uint32_t abort_isp_active:1;		/* 4 */
 		uint32_t disable_risc_code_load:1;	/* 5 */
-#ifdef __ia64__
+#ifdef CONFIG_IA64_SGI_SN2
 		uint32_t use_pci_vchannel:1;
 #endif
 	} flags;
diff --git a/drivers/scsi/qla2xxx/qla_init.c b/drivers/scsi/qla2xxx/qla_init.c
index 27f22ca7b038..32d14c414344 100644
--- a/drivers/scsi/qla2xxx/qla_init.c
+++ b/drivers/scsi/qla2xxx/qla_init.c
@@ -16,7 +16,7 @@
 #include <asm/prom.h>
 #endif
 
-#ifdef CONFIG_IA64
+#ifdef CONFIG_IA64_SGI_SN2
 #include <asm/machvec.h>
 #endif
 
